File: devenv~/RuntimeTypescriptTests/tests/test-method-with-a-lot-of-arguments.ts
import { UnityInstance } from "../UnityInstance.js";
import { assertEquals, SampleValues } from "./test-utilities.js";


/**
 * Test ensuring that methods can be called up to the limit of 10 arguments of the library
 */
export function RunTests(unityInstance: UnityInstance) {

  // Get module
  const module = unityInstance.Module;
  const testStaticClass = module.static["Nahoum.UnityJSInterop.Tests"].TestMethodWithALotOfArguments;
  const testInt = module.utilities.GetManagedInt(1);

  // Test 0 arg
  assertEquals(testStaticClass.OneArgument(testInt).value, 1, "Test0Args");
  assertEquals(testStaticClass.TwoArguments(testInt, testInt).value, 2, "Test1Arg");
  assertEquals(testStaticClass.ThreeArguments(testInt, testInt, testInt).value, 3, "Test2Args");
  assertEquals(testStaticClass.FourArguments(testInt, testInt, testInt, testInt).value, 4, "Test3Args");
  assertEquals(testStaticClass.FiveArguments(testInt, testInt, testInt, testInt, testInt).value, 5, "Test4Args");
  assertEquals(testStaticClass.SixArguments(testInt, testInt, testInt, testInt, testInt, testInt).value, 6, "Test5Args");
  assertEquals(testStaticClass.SevenArguments(testInt, testInt, testInt, testInt, testInt, testInt, testInt).value, 7, "Test6Args");
  assertEquals(testStaticClass.EightArguments(testInt, testInt, testInt, testInt, testInt, testInt, testInt, testInt).value, 8, "Test7Args");
  assertEquals(testStaticClass.NineArguments(testInt, testInt, testInt, testInt, testInt, testInt, testInt, testInt, testInt).value, 9, "Test8Args");
  assertEquals(testStaticClass.TenArguments(testInt, testInt, testInt, testInt, testInt, testInt, testInt, testInt, testInt, testInt).value, 10, "Test9Args");
}

File: devenv~/Assets/Tests/TestMethodWithALotOfArguments.cs
using AnotherNamespace;
using Nahoum.UnityJSInterop;
using Nahoum.UnityJSInterop.Tests;

namespace Nahoum.UnityJSInterop.Tests
{
    /// <summary>
    /// Test the maximum number of arguments we may have in exposed methods.
    /// This is limited by the registry.
    /// </summary>
    public class TestMethodWithALotOfArguments
    {
        /// <summary>
        /// Method with 1 argument.
        /// </summary>
        [ExposeWeb]
        public static int OneArgument(int firstArg)
        {
            return firstArg;
        }

        /// <summary>
        /// Method with 2 arguments.
        /// </summary>
        [ExposeWeb]
        public static int TwoArguments(int firstArg, int secondArg)
        {
            return Sum(firstArg, secondArg);
        }

        /// <summary>
        /// Method with 3 arguments.
        /// </summary>
        [ExposeWeb]
        public static int ThreeArguments(int firstArg, int secondArg, int thirdArg)
        {
            return Sum(firstArg, secondArg, thirdArg);
        }

        /// <summary>
        /// Method with 4 arguments.
        /// </summary>
        [ExposeWeb]
        public static int FourArguments(int firstArg, int secondArg, int thirdArg, int fourthArg)
        {
            return Sum(firstArg, secondArg, thirdArg, fourthArg);
        }

        /// <summary>
        /// Method with 5 arguments.
        /// </summary>
        [ExposeWeb]
        public static int FiveArguments(int firstArg, int secondArg, int thirdArg, int fourthArg, int fifthArg)
        {
            return Sum(firstArg, secondArg, thirdArg, fourthArg, fifthArg);
        }

        /// <summary>
        /// Method with 6 arguments.
        /// </summary>
        [ExposeWeb]
        public static int SixArguments(int firstArg, int secondArg, int thirdArg, int fourthArg, int fifthArg, int sixthArg)
        {
            return Sum(firstArg, secondArg, thirdArg, fourthArg, fifthArg, sixthArg);
        }

        /// <summary>
        /// Method with 7 arguments.
        /// </summary>
        [ExposeWeb]
        public static int SevenArguments(int firstArg, int secondArg, int thirdArg, int fourthArg, int fifthArg, int sixthArg, int seventhArg)
        {
            return Sum(firstArg, secondArg, thirdArg, fourthArg, fifthArg, sixthArg, seventhArg);
        }

        /// <summary>
        /// Method with 8 arguments.
        /// </summary>
        [ExposeWeb]
        public static int EightArguments(int firstArg, int secondArg, int thirdArg, int fourthArg, int fifthArg, int sixthArg, int seventhArg, int eighthArg)
        {
            return Sum(firstArg, secondArg, thirdArg, fourthArg, fifthArg, sixthArg, seventhArg, eighthArg);
        }

        /// <summary>
        /// Method with 9 arguments.
        /// </summary>
        [ExposeWeb]
        public static int NineArguments(int firstArg, int secondArg, int thirdArg, int fourthArg, int fifthArg, int sixthArg, int seventhArg, int eighthArg, int ninthArg)
        {
            return Sum(firstArg, secondArg, thirdArg, fourthArg, fifthArg, sixthArg, seventhArg, eighthArg, ninthArg);
        }

        /// <summary>
        /// Method with 10 arguments.
        /// </summary>
        [ExposeWeb]
        public static int TenArguments(int firstArg, int secondArg, int thirdArg, int fourthArg, int fifthArg, int sixthArg, int seventhArg, int eighthArg, int ninthArg, int tenthArg)
        {
            return Sum(firstArg, secondArg, thirdArg, fourthArg, fifthArg, sixthArg, seventhArg, eighthArg, ninthArg, tenthArg);
        }

        /// <summary>
        /// Helper method that sums an array of integers.
        /// </summary>
        private static int Sum(params int[] args)
        {
            int sum = 0;
            for (int i = 0; i < args.Length; i++)
            {
                sum += args[i];
            }
            return sum;
        }
    }
}
